Aplib.Core/DataStructures/OptimizedActivationStack.cs

```diff
@@ -0,0 +1,222 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace Aplib.Core.DataStructures
+{
+    /// <summary>
+    /// A stack that has a predefined set of items that can be <i>activated</i>
+    /// (i.e., pushed on top of the stack).
+    /// When an item that is already on the stack is activated, it is <i>reactivated</i>
+    /// (i.e., moved to the top of the stack).
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="OptimizedActivationStack{T}"/> allows for O(1) activation and
+    /// reactivation of an arbitrary stack item.
+    /// </remarks>
+    /// <typeparam name="T">The type of the items that are put on the stack.</typeparam>
+    public class OptimizedActivationStack<T>
+    {
+        /// <summary>
+        /// The top item on the stack.
+        /// </summary>
+        private Item? _top;
+
+        /// <summary>
+        /// Gets the activatable stack items.
+        /// </summary>
+        /// <remarks>
+        /// The stack items are exposed, since they should be accessible from the outside to
+        /// provide O(1) activation of a stack item with <see cref="Activate(Item)"/>.
+        /// </remarks>
+        public IEnumerable<Item> ActivatableStackItems { get; }
+
+        /// <summary>
+        /// Gets the number of items that are currently activated (i.e., on the stack).
+        /// </summary>
+        public int Count { get; private set; } = 0;
+
+        /// <summary>
+        /// Initializes an optimized activation stack with a set of activatable data.
+        /// </summary>
+        /// <param name="activatables">A set of activatable items that could be pushed on the stack.</param>
+        public OptimizedActivationStack(T[] activatables)
+        {
+            // Setup the activatable stack items.
+            ActivatableStackItems = activatables.Select(activatable => new Item(activatable, this));
+        }
+
+        /// <summary>
+        /// Activates an item (i.e., pushes an item on top of the stack).
+        /// If the pushed item is already on the stack,
+        /// it is extracted from the stack before it is put on top again.
+        /// </summary>
+        /// <param name="item">
+        /// The stack item that is pushed on top of the stack (i.e., it is activated).
+        /// </param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when an item is pushed that belongs to a different stack.
+        /// </exception>
+        public void Activate(Item item)
+        {
+            if (item.ActivationStack != this)
+                throw new System.ArgumentException("Cannot push an item that is not an activatable of this activation stack.");
+
+            // Handle the case when the stack is empty.
+            if (_top is null)
```

Comment:
Wouldn't it be better semantically to test for `Count <= 0`?

```suggestion
            if (Count <= 0)
```

Reply:
Great idea